metaflow/plugins/kfp/kfp.py

```diff
@@ -1,86 +1,275 @@
+
+    StepOutput = namedtuple('StepOutput',
+                             ['ds_root', 'run_id'])
+
+    # TODO: Check why MF echo statements are getting redirected to stderr
+    if len(proc_error) > 1:
+        print("_____________ STDERR:_____________________________")
+        print(proc_error)
+
+    if len(proc_output) > 1:
+        print("______________ STDOUT:____________________________")
+        print(proc_output)
+
+        # Read in the outputs (in the penultimate line) containing args needed for the next steps.
+        # Note: Output format is: ['ds_root', 'run_id']
+        outputs = (proc_output.split("\n")[-2]).split() # this contains the args needed for next steps to run
+
+    else:
+        raise RuntimeWarning("This step did not generate the correct args for next step to run. This might disrupt the workflow")
+
+    # TODO: Metadata needed for client API to run needs to be persisted outside before this
+    print("--------------- RUNNING: CLEANUP OF TEMP DIR----------")
+    subprocess.call(["rm -r /opt/zillow/.metaflow"], shell=True)
+
+    print("_______________ Done __________________________")
+    return StepOutput(outputs[0], outputs[1])
+
+def step_container_op():
     """
-    Function that creates the KFP flow pipeline and returns the path to the YAML file containing the pipeline specification.
+    Container op that corresponds to a step defined in the Metaflow flowgraph.
+
+    Note: The public docker image is a copy of the internal docker image we were using (borrowed from aip-kfp-example).
     """
 
-    steps = ordered_steps
+    step_op = kfp.components.func_to_container_op(step_op_func, base_image='ssreejith3/mf_on_kfp:python-curl-git')
+    return step_op
+
+def pre_start_container_op():
+    """
+    Container op that corresponds to the 'pre-start' step of Metaflow.
+
+    Note: The public docker image is a copy of the internal docker image we were using (borrowed from aip-kfp-example).
+    """
+
+    pre_start_op = kfp.components.func_to_container_op(pre_start_op_func, base_image='ssreejith3/mf_on_kfp:python-curl-git')
+    return pre_start_op
+
+def create_command_templates_from_graph(graph):
+    """
+    Create a map of steps to their corresponding command templates. These command templates help define the command
+    to be used to run that particular step with placeholders for the `run_id` and `datastore_root` (location of the datastore).
+
+    # Note:
+    # Level-order traversal is adopted to keep the task-ids in line with what happens during a local metaflow execution.
+    # It is not entirely necessary to keep this order of task-ids if we are able to point to the correct input-paths for
+    # each step. But, using this ordering does keep the organization of data more understandable and natural (i.e., `start`
+    # gets a task id of 1, next step gets a task id of 2 and so on with 'end' step having the highest task id.
+    """
+    def build_cmd_template(step_name, task_id, input_paths):
```

Comment:
Would be mind adding in a quick docstring here? Not having type hints (which I understand is the MF style) is making it pretty tough to review at times!

Reply:
Done. I've added an example of what this template looks like as well. Thanks for pointing this out :)

Reply:
oh yes, and please add docstrings to all functions.  A comment in one place, would apply to all, for an consistent codebase.